### PR TITLE
Add spelling to pylint checks

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |
| ✓   | :scroll: Docs          |

## Description

This add our own spelling checker to the pylint's check. Introduce a dictionary.

### CODE REVIEW

File: pylint/reporters/text.py [hunk 1]
```diff
@@ -71,11 +71,11 @@ class MessageStyle(NamedTuple):
 
 
 def _get_ansi_code(msg_style: MessageStyle) -> str:
-    """Return ansi escape code corresponding to color and style.
+    """Return ANSI escape code corresponding to color and style.
 
     :param msg_style: the message style
 
-    :raise KeyError: if an unexistent color or style identifier is given
+    :raise KeyError: if a non-existent color or style identifier is given
```
Comment (jacobtylerwalls) [inline pylint/reporters/text.py:78]:
```suggestion
    :raise KeyError: if a nonexistent color or style identifier is given
```

File: pylint/checkers/typecheck.py [hunk 2]
```diff
@@ -222,7 +222,7 @@ def _missing_member_hint(owner, attrname, distance_threshold, max_choices):
     "E1101": (
         "%s %r has no %r member%s",
         "no-member",
-        "Used when a variable is accessed for an unexistent member.",
+        "Used when a variable is accessed for a non-existent member.",
```
Comment (jacobtylerwalls) [inline pylint/checkers/typecheck.py:225]:
```suggestion
        "Used when a variable is accessed for a nonexistent member.",
```

File: ChangeLog [hunk 3]
```diff
@@ -6697,7 +6697,7 @@ Release date: 2006-03-06
 
     * a new typecheck checker, introducing the following checks:
 
-        - E1101, access to unexistent member (implements #10430), remove
+        - E1101, access to non-existent member (implements #10430), remove
```
Comment (jacobtylerwalls) [inline ChangeLog:6700]:
```suggestion
        - E1101, access to nonexistent member (implements #10430), remove
```

File: examples/custom.py [hunk 4]
```diff
@@ -15,7 +15,7 @@ class MyAstroidChecker(BaseChecker):
 
     # The name defines a custom section of the config for this checker.
     name = "custom"
-    # This class variable declares the messages (ie the warnings and errors)
+    # This class variable declares the messages (i.e.the warnings and errors)
```
Comment (jacobtylerwalls) [inline examples/custom.py:18]:
```suggestion
    # This class variable declares the messages (i.e. the warnings and errors)
```

File: pylint/checkers/variables.py [hunk 5]
```diff
@@ -2585,7 +2585,7 @@ def _check_unpacking(self, inferred, node, targets):
                         len(values),
                     ),
                 )
-        # attempt to check unpacking may be possible (ie RHS is iterable)
+        # attempt to check unpacking may be possible (i.e.RHS is iterable)
```
Comment (jacobtylerwalls) [inline pylint/checkers/variables.py:2588]:
```suggestion
        # attempt to check unpacking may be possible (i.e. RHS is iterable)
```

File: tests/test_self.py [hunk 6]
```diff
@@ -564,8 +568,7 @@ def foobar(arg):
                     expected_output=expected,
                 )
 
-            # this code needs to work w/ and w/o a file named a/b.py on the
-            # harddisk.
+            # this code needs to work w/ and w/o a file named a/b.py on the hard-disk.
```
Comment (jacobtylerwalls) [inline tests/test_self.py:571]:
```suggestion
            # this code needs to work w/ and w/o a file named a/b.py on the hard disk.
```

File: ChangeLog [hunk 7]
```diff
@@ -3191,7 +3191,7 @@ Release date: 2019-09-24
 
 * Allow a ``.`` as a prefix for Sphinx name resolution.
 
-* Checkers must now keep a 1 to 1 relationship between "msgid" (ie: C1234) and "symbol" (ie : human-readable-symbol)
+* Checkers must now keep a 1 to 1 relationship between "msgid" (ie: C1234) and "symbol" (i.e.: human-readable-symbol)
```
Comment (jacobtylerwalls) [inline ChangeLog:3194]:
```suggestion
* Checkers must now keep a 1 to 1 relationship between "msgid" (ie: C1234) and "symbol" (i.e. : human-readable-symbol)
```

**Review by jacobtylerwalls — COMMENTED** [review]
Nice win for the review process.

File: tests/test_functional.py [hunk 8]
```diff
@@ -22,10 +22,6 @@
 )
 from pylint.utils import HAS_ISORT_5
```
Comment (DanielNoord) [inline tests/test_functional.py:24]: ?

File: tests/test_self.py [hunk 9]
```diff
@@ -265,8 +265,12 @@ def test_enable_all_works(self) -> None:
         {module}:22:0: C0115: Missing class docstring (missing-class-docstring)
         """
         )
+        # Use an empty pylintrc, otherwise, we'd use the project's pylintrc
+        # which is less stable than the default values.
+        pylintrc = os.path.join(HERE, "data", "pylintrc")
         self._test_output(
-            [module, "--disable=all", "--enable=all", "-rn"], expected_output=expected
+            [module, "--disable=all", "--enable=all", "-rn", f"--rcfile={pylintrc}"],
```
Comment (DanielNoord) [inline tests/test_self.py:272]: This should ideally be done in a separate PR.

File: .pyenchant_pylint_custom_dict.txt [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+bla
+bom
+bool
+boolean
+boolop
+boundmethod
+builtins
+bw
+callables
+cardinality
```
Comment (DanielNoord) [inline .pyenchant_pylint_custom_dict.txt:40]: This makes me wonder how up to date the base dictionary is..

Comment (DanielNoord) [reply .pyenchant_pylint_custom_dict.txt:40]: Is there a `EN_US/EN_GB` dictionary? That combines the two?

File: tests/test_functional.py (hunk 8 above)
Comment (Pierre-Sassoulas) [reply tests/test_functional.py:24]: I never understood what "exhaustivity tests" means, so the TODO is inapplicable, and I think it might be the ``--enable=one-message --disable=all`` done by Andreas recentely.

File: .pyenchant_pylint_custom_dict.txt (hunk 10 above)
Comment (Pierre-Sassoulas) [reply .pyenchant_pylint_custom_dict.txt:40]: Turns out that there might be a simple solution : https://stackoverflow.com/a/69540620/2519059

File: tests/test_self.py (hunk 9 above)
Comment (Pierre-Sassoulas) [reply tests/test_self.py:272]: Turn out the separate PR was required as it took me 4/5 hours to do correctly. https://github.com/PyCQA/pylint/pull/6459

File: .pyenchant_pylint_custom_dict.txt (hunk 10 above)
Comment (Pierre-Sassoulas) [reply .pyenchant_pylint_custom_dict.txt:40]:
```diff
diff --git a/.pyenchant_pylint_custom_dict.txt b/.pyenchant_pylint_custom_dict.txt
index e4a1f8f06..67086db8c 100644
--- a/.pyenchant_pylint_custom_dict.txt
+++ b/.pyenchant_pylint_custom_dict.txt
@@ -26,7 +26,6 @@ backend
 backticks
 BaseChecker
 basename
-behaviour
 bidi
 bla
 bom
diff --git a/pylintrc b/pylintrc
index dbe919b47..c45d9f38d 100644
--- a/pylintrc
+++ b/pylintrc
@@ -392,7 +392,7 @@ missing-member-max-choices=1
 
 # Spelling dictionary name. Available dictionaries: none. To make it working
 # install python-enchant package.
-spelling-dict=en_US
+spelling-dict=en
 
 # List of comma separated words that should not be checked.
 spelling-ignore-words=
```
This does not work sadly. I think we'll keep our little private dict up to date it's the simple way to fix this.

Comment (DanielNoord) [reply .pyenchant_pylint_custom_dict.txt:40]: Hm, I was reading the read me of pyenchant and it actually says it's quite flexible in handling multiple languages.

I'd be willing to spend some time to look whether we can make the dictionary option a csv (which I think is backwards compatible) and then allow the usage of multiple.

Comment (Pierre-Sassoulas) [reply .pyenchant_pylint_custom_dict.txt:40]: > I'd be willing to spend some time to look whether we can make the dictionary option a csv (which I think is backwards compatible) and then allow the usage of multiple.

I'll check the windows tests failing, and there's a lot of MR in the pipe right now feel free to update this one :)

Comment (Pierre-Sassoulas) [reply .pyenchant_pylint_custom_dict.txt:40]: Well nvm, the windows test magically got fixed, probably thanks to #6459. Would you mind doing a follow-up ? Because every MR that is not spell checked add new issues.

**Review by DanielNoord — APPROVED** [review]

**Comment by DanielNoord** [comment]
Let's revert the change to our `pylintrc`. This breaks `main` as you now need the `enchant` package to run `pylint pylint` after doing `pip install -e .`.
See https://pyenchant.github.io/pyenchant/install.html#installation.

We can probably add a step in our CI where we run with `pylint --spelling-dict=en_US` but it shouldn't be part of our `pylintrc`.

**Comment by Pierre-Sassoulas** [comment]
My bad, I opened https://github.com/PyCQA/pylint/pull/6462
